synapse/handlers/e2e_room_keys.py

```diff
@@ -166,43 +182,43 @@ def upload_room_keys(self, user_id, version, room_keys):
                     else:
                         raise
 
-            # go through the room_keys.
-            # XXX: this should/could be done concurrently, given we're in a lock.
+            # Fetch any existing room keys for the sessions that have been
+            # submitted.  Then compare them with the submitted keys.  If the
+            # key is new, insert it; if the key should be updated, then update
+            # it; otherwise, drop it.
+            existing_keys = yield self.store.get_e2e_room_keys_multi(
+                user_id, version, room_keys["rooms"]
+            )
+            to_insert = []  # batch the inserts together
+            changed = False  # if anything has changed, we need to update the hash
             for room_id, room in iteritems(room_keys["rooms"]):
-                for session_id, session in iteritems(room["sessions"]):
```

Comment:
FWIW, the diff gets hard to follow in this section, and it's much easier to just [look at the file](https://github.com/matrix-org/synapse/blob/d081883f1e2bffabbcbc94d352ca819b19fd32fd/synapse/handlers/e2e_room_keys.py#L194-L221) for this section.